File: python/merge_sort.py
import sys


def merge(left: list, right: list) -> list:
    merged = []
    left_index = 0
    right_index = 0

    while left_index < len(left) and right_index < len(right):
        if left[left_index] <= right[right_index]:
            merged.append(left[left_index])
            left_index += 1
        else:
            merged.append(right[right_index])
            right_index += 1

    if left_index < len(left):
        merged.extend(left[left_index:])
    elif right_index < len(right):
        merged.extend(right[right_index:])

    return merged


def merge_sort(data: list) -> list:
    if len(data) <= 1:
        return data

    mid = len(data) // 2
    left = merge_sort(data[:mid])
    right = merge_sort(data[mid:])

    return merge(left, right)


def main(file_path: str, write_output: bool = False):
    with open(file_path, 'r') as f:
        int(f.readline())
        data = f.read().split(';')
        data = list(map(int, data))

    sorted = merge_sort(data)

    if write_output:
        with open('result.txt', 'w') as f:
            f.write(str(len(sorted)) + '\n')
            f.write(';'.join(map(str, sorted)))


if __name__ == '__main__':
    # Read file given as first argument and split it into values separated by ;

    if len(sys.argv) < 2 or len(sys.argv) > 3:
        print('Usage: python merge_sort.py <file> [write_output]')
        sys.exit(1)

    if len(sys.argv) == 3:
        write_output = sys.argv[2] == 'true'
    else:
        write_output = False

    if not sys.argv[1].endswith('.txt'):
        print('File must be a .txt file')
        sys.exit(1)

    main(sys.argv[1], write_output)

File: java/MergeSort.java
import java.io.*;
import java.util.Arrays;

public class MergeSort {

    public static int[] merge(int[] left, int[] right) {
        int[] merged = new int[left.length + right.length];
        int leftIndex = 0, rightIndex = 0, mergeIndex = 0;

        while (leftIndex < left.length && rightIndex < right.length) {
            if (left[leftIndex] <= right[rightIndex]) {
                merged[mergeIndex++] = left[leftIndex++];
            } else {
                merged[mergeIndex++] = right[rightIndex++];
            }
        }

        while (leftIndex < left.length) {
            merged[mergeIndex++] = left[leftIndex++];
        }

        while (rightIndex < right.length) {
            merged[mergeIndex++] = right[rightIndex++];
        }

        return merged;
    }

    public static int[] mergeSort(int[] data) {
        if (data.length <= 1) {
            return data;
        }

        int mid = data.length / 2;
        int[] left = Arrays.copyOfRange(data, 0, mid);
        int[] right = Arrays.copyOfRange(data, mid, data.length);

        left = mergeSort(left);
        right = mergeSort(right);

        return merge(left, right);
    }

    public static void main(String[] args) {
        if (args.length < 1 || args.length > 2) {
            System.out.println("Usage: java MergeSort <file> [write_output]");
            System.exit(1);
        }

        String filePath = args[0];
        boolean writeOutput = args.length == 2 && args[1].equals("true");

        if (!filePath.endsWith(".txt")) {
            System.out.println("File must be a .txt file");
            System.exit(1);
        }

        try {
            BufferedReader reader = new BufferedReader(new FileReader(filePath));
            reader.readLine(); // Skip first line
            String[] data = reader.readLine().split(";");
            int[] intData = new int[data.length];
            for (int i = 0; i < data.length; i++) {
                intData[i] = Integer.parseInt(data[i].trim());
            }
            reader.close();

            int[] sorted = mergeSort(intData);

            if (writeOutput) {
                BufferedWriter writer = new BufferedWriter(new FileWriter("result.txt"));
                writer.write(sorted.length + "\n");
                for (int i = 0; i < sorted.length; i++) {
                    writer.write(sorted[i] + (i < sorted.length - 1 ? ";" : ""));
                }
                writer.close();
            }

        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}
